### PR TITLE
Adds SQL support for Configurable Table Snapshot History

### PR DESCRIPTION
**Pull request by Will-Lo**
## Summary

Adds Spark SQL support for configurable table snapshots, which controls the versioning of the Openhouse tables.
Syntax is similar to retention but is instead defined as `HISTORY`.

History configuration supports both `MAX_AGE` and `VERSIONS`, where we retain all table snapshots that live within `MAX_AGE` and within `VERSIONS`.

Example: A table with `MAX_AGE = 1d` will retain all snapshots that are within 1 day of when the snapshot retention job last ran.
A table with `VERSIONS = 5` will retain the last 5 snapshots of the table without considering the age of the snapshots
If both `MAX_AGE = 1d` and `VERSIONS = 5` is defined, keep the last 5 snapshots within the last day. Note: If there are less than 5 snapshots, then there were less than 5 commits done in the past day.

`MAX_AGE` and `VERSIONS` cannot be defined as less than 1.
The default maximums of `MAX_AGE` and `VERSIONS` defined in https://github.com/linkedin/openhouse/pull/259 are 3 days and 100 versions respectively.

Examples:
```
ALTER TABLE <db>.<table> SET POLICY (HISTORY MAX_AGE=24H VERSIONS=10)
ALTER TABLE <db>.<table> SET POLICY (HISTORY MAX_AGE=3D)
ALTER TABLE <db>.<table> SET POLICY (HISTORY VERSIONS=5)
```

## Changes

- [x] Client-facing API Changes
- [x] Internal API Changes
- [ ] Bug Fixes
- [x] New Features
- [ ] Performance Improvements
- [ ] Code Style
- [ ] Refactoring
- [ ] Documentation
- [x] Tests

For all the boxes checked, please include additional details of the changes made in this pull request.  

## Testing Done

- [x] Manually Tested on local docker setup. Please include commands ran, and their output.
- [x] Added new tests for the changes made.
- [x] Updated existing tests to reflect the changes made.
- [ ] No tests added or updated. Please explain why. If unsure, please feel free to ask for help.
- [ ] Some other form of testing like staging or soak time in production. Please explain.

Tested on local docker running spark:
Tested setting both policies
```
scala> spark.sql("ALTER TABLE openhouse.db.tb2 SET POLICY ( HISTORY MAX_AGE=3d VERSIONS=3 )").show
++
||
++
++


scala> spark.sql("SHOW TBLPROPERTIES openhouse.db.tb2 (policies)").show(truncate=false)
+--------+-----------------------------------------------------------------------------------------------------------------+
|key     |value                                                                                                            |
+--------+-----------------------------------------------------------------------------------------------------------------+
|policies|{
  "sharingEnabled": false,
  "history": {
    "maxAge": 3,
    "granularity": "DAY",
    "versions": 3
  }
}|
+--------+-----------------------------------------------------------------------------------------------------------------+

```

Setting only versions:
```
scala> spark.sql("ALTER TABLE openhouse.db.tb2 SET POLICY ( HISTORY VERSIONS=20 )").show
++
||
++
++


scala> spark.sql("SHOW TBLPROPERTIES openhouse.db.tb2 (policies)").show(truncate=false)
+--------+----------------------------------------------------------------------------------------+
|key     |value                                                                                   |
+--------+----------------------------------------------------------------------------------------+
|policies|{
  "sharingEnabled": false,
  "history": {
    "maxAge": 0,
    "versions": 20
  }
}|
+--------+----------------------------------------------------------------------------------------+
```

Setting only max age
```
scala> spark.sql("ALTER TABLE openhouse.db.tb2 SET POLICY ( HISTORY MAX_AGE=8h )").show
++
||
++
++


scala> spark.sql("SHOW TBLPROPERTIES openhouse.db.tb2 (policies)").show(truncate=false)
+--------+------------------------------------------------------------------------------------------------------------------+
|key     |value                                                                                                             |
+--------+------------------------------------------------------------------------------------------------------------------+
|policies|{
  "sharingEnabled": false,
  "history": {
    "maxAge": 8,
    "granularity": "HOUR",
    "versions": 0
  }
}|
+--------+------------------------------------------------------------------------------------------------------------------+
```

Also tested negative cases (invalid numbers, past maximums defined in https://github.com/linkedin/openhouse/pull/259)
e.g.

```
scala> spark.sql("ALTER TABLE openhouse.db.tb SET POLICY ( HISTORY MAX_AGE=1h VERSIONS=2 )").show
org.apache.iceberg.exceptions.BadRequestException: 400 , {"status":"BAD_REQUEST","error":"Bad Request","message":" : History for the table [LocalHadoopCluster.db.tb] max age must be between 1 to 3 days","stacktrace":null,"cause":"Not Available"}


```

For all the boxes checked, include a detailed description of the testing done for the changes made in this pull request.

# Additional Information

- [ ] Breaking Changes
- [ ] Deprecations
- [ ] Large PR broken into smaller PRs, and PR plan linked in the description.

For all the boxes checked, include additional details of the changes made in this pull request.

### CODE REVIEW

File: integrations/spark/spark-3.1/openhouse-spark-itest/src/test/java/com/linkedin/openhouse/spark/e2e/extensions/SetSnapshotsRetentionPolicyTest.java [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+            "dbSetSnapshotsRetention.t3",
+            "u1",
+            mockTableLocation(
+                TableIdentifier.of("dbSetSnapshotsRetention", "t3"),
+                convertSchemaToDDLComponent(baseSchema),
+                ""),
+            "V1",
+            baseSchema,
+            null,
+            null);
```
Comment (abhisheknath2011) [inline integrations/spark/spark-3.1/openhouse-spark-itest/src/test/java/com/linkedin/openhouse/spark/e2e/extensions/SetSnapshotsRetentionPolicyTest.java:138]: As this is repeated multiple times can we initialize the table in separate method?

File: integrations/spark/spark-3.1/openhouse-spark-itest/src/test/java/com/linkedin/openhouse/spark/e2e/extensions/SetSnapshotsRetentionPolicyTest.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    mockTableService.enqueue(mockResponse(200, existingOpenhouseTable)); // doRefresh()
+    mockTableService.enqueue(mockResponse(200, existingOpenhouseTable)); // doRefresh()
+    mockTableService.enqueue(mockResponse(201, tableAfterSetSnapshotsRetention)); // doCommit()
+    String ddlWithSchema =
+        "ALTER TABLE openhouse.dbSetSnapshotsRetention.t3 SET POLICY (VERSIONS TIME=24H)";
+    Assertions.assertDoesNotThrow(() -> spark.sql(ddlWithSchema));
+  }
+
+  @Test
+  public void testSetSnapshotsRetentionPolicyCountOnly() {
```
Comment (abhisheknath2011) [inline integrations/spark/spark-3.1/openhouse-spark-itest/src/test/java/com/linkedin/openhouse/spark/e2e/extensions/SetSnapshotsRetentionPolicyTest.java:177]: Can we add some negative tests as well?

File: integrations/spark/spark-3.1/openhouse-spark-runtime/src/main/scala/com/linkedin/openhouse/spark/sql/catalyst/parser/extensions/OpenhouseSqlExtensionsAstBuilder.scala [hunk 3]
```diff
@@ -149,7 +150,7 @@ class OpenhouseSqlExtensionsAstBuilder (delegate: ParserInterface) extends Openh
   }
 
   override def visitDuration(ctx: DurationContext): (String, Int) = {
-    val granularity = if (ctx.RETENTION_DAY != null) {
+    val granularity: String = if (ctx.RETENTION_DAY != null) {
```
Comment (abhisheknath2011) [inline integrations/spark/spark-3.1/openhouse-spark-runtime/src/main/scala/com/linkedin/openhouse/spark/sql/catalyst/parser/extensions/OpenhouseSqlExtensionsAstBuilder.scala:153]: What level of granularity is supported like hour and day only? What if someone provides months, is there a validation not to support that?

**Review by abhisheknath2011 — COMMENTED** [review]
Thanks @Will-Lo. Overall looks good. Added some comments.

File: integrations/spark/spark-3.1/openhouse-spark-itest/src/test/java/com/linkedin/openhouse/spark/e2e/extensions/SetSnapshotsRetentionPolicyTest.java (hunk 1 above)
Comment (Will-Lo) [reply integrations/spark/spark-3.1/openhouse-spark-itest/src/test/java/com/linkedin/openhouse/spark/e2e/extensions/SetSnapshotsRetentionPolicyTest.java]: Moved tests to `SetHistoryPolicyStatementTest` which doesn't need this bootstrap

File: integrations/spark/spark-3.1/openhouse-spark-itest/src/test/java/com/linkedin/openhouse/spark/e2e/extensions/SetSnapshotsRetentionPolicyTest.java (hunk 2 above)
Comment (Will-Lo) [reply integrations/spark/spark-3.1/openhouse-spark-itest/src/test/java/com/linkedin/openhouse/spark/e2e/extensions/SetSnapshotsRetentionPolicyTest.java]: Done in `SetHistoryPolicyStatementTest`

File: integrations/spark/spark-3.1/openhouse-spark-runtime/src/main/scala/com/linkedin/openhouse/spark/sql/catalyst/parser/extensions/OpenhouseSqlExtensionsAstBuilder.scala (hunk 3 above)
Comment (Will-Lo) [reply integrations/spark/spark-3.1/openhouse-spark-runtime/src/main/scala/com/linkedin/openhouse/spark/sql/catalyst/parser/extensions/OpenhouseSqlExtensionsAstBuilder.scala:153]: Maximums is going to be validated at the API level I'm thinking: https://github.com/linkedin/openhouse/pull/259

**Comment by maluchari** [comment]
> ## Summary
> Adds Spark SQL support for configurable table snapshots, which controls the versioning of the Openhouse tables. Syntax is similar to retention but is instead defined as `HISTORY`.
> 
> Examples:
> 
> ```
> ALTER TABLE <db>.<table> SET POLICY (HISTORY MAX_AGE=24H MIN_VERSIONS=10)
> ALTER TABLE <db>.<table> SET POLICY (HISTORY MAX_AGE=3D)
> ALTER TABLE <db>.<table> SET POLICY (HISTORY MIN_VERSIONS=5)
> ```
> 
> ## Changes
> * [x]  Client-facing API Changes
> * [x]  Internal API Changes
> * [ ]  Bug Fixes
> * [x]  New Features
> * [ ]  Performance Improvements
> * [ ]  Code Style
> * [ ]  Refactoring
> * [ ]  Documentation
> * [x]  Tests
> 
> For all the boxes checked, please include additional details of the changes made in this pull request.
> 
> ## Testing Done
> * [x]  Manually Tested on local docker setup. Please include commands ran, and their output.
> * [x]  Added new tests for the changes made.
> * [x]  Updated existing tests to reflect the changes made.
> * [ ]  No tests added or updated. Please explain why. If unsure, please feel free to ask for help.
> * [ ]  Some other form of testing like staging or soak time in production. Please explain.
> 
> Tested on local docker running spark: Tested setting both policies
> 
> ```
> scala> spark.sql("ALTER TABLE openhouse.db.tb2 SET POLICY ( HISTORY MAX_AGE=3d MIN_VERSIONS=3 )").show
> ++
> ||
> ++
> ++
> 
> 
> scala> spark.sql("SHOW TBLPROPERTIES openhouse.db.tb2 (policies)").show(truncate=false)
> +--------+-----------------------------------------------------------------------------------------------------------------+
> |key     |value                                                                                                            |
> +--------+-----------------------------------------------------------------------------------------------------------------+
> |policies|{
>   "sharingEnabled": false,
>   "history": {
>     "maxAge": 3,
>     "granularity": "DAY",
>     "minVersions": 3
>   }
> }|
> +--------+-----------------------------------------------------------------------------------------------------------------+
> ```
> 
> Setting only versions:
> 
> ```
> scala> spark.sql("ALTER TABLE openhouse.db.tb2 SET POLICY ( HISTORY MIN_VERSIONS=20 )").show
> ++
> ||
> ++
> ++
> 
> 
> scala> spark.sql("SHOW TBLPROPERTIES openhouse.db.tb2 (policies)").show(truncate=false)
> +--------+----------------------------------------------------------------------------------------+
> |key     |value                                                                                   |
> +--------+----------------------------------------------------------------------------------------+
> |policies|{
>   "sharingEnabled": false,
>   "history": {
>     "maxAge": 0,
>     "minVersions": 20
>   }
> }|
> +--------+----------------------------------------------------------------------------------------+
> ```
> 
> Setting only max age
> 
> ```
> scala> spark.sql("ALTER TABLE openhouse.db.tb2 SET POLICY ( HISTORY MAX_AGE=8h )").show
> ++
> ||
> ++
> ++
> 
> 
> scala> spark.sql("SHOW TBLPROPERTIES openhouse.db.tb2 (policies)").show(truncate=false)
> +--------+------------------------------------------------------------------------------------------------------------------+
> |key     |value                                                                                                             |
> +--------+------------------------------------------------------------------------------------------------------------------+
> |policies|{
>   "sharingEnabled": false,
>   "history": {
>     "maxAge": 8,
>     "granularity": "HOUR",
>     "minVersions": 0
>   }
> }|
> +--------+------------------------------------------------------------------------------------------------------------------+
> ```
> 
> Also tested negative cases (invalid numbers, past maximums defined in #259)
> 
> For all the boxes checked, include a detailed description of the testing done for the changes made in this pull request.
> 
> # Additional Information
> * [ ]  Breaking Changes
> * [ ]  Deprecations
> * [ ]  Large PR broken into smaller PRs, and PR plan linked in the description.
> 
> For all the boxes checked, include additional details of the changes made in this pull request.

Can we please add details regarding the SQL API contract? If both versions and max age is present what is expected for example.

File: integrations/spark/spark-3.1/openhouse-spark-runtime/src/main/antlr/com/linkedin/openhouse/spark/sql/catalyst/parser/extensions/OpenhouseSqlExtensions.g4 [hunk 4]
```diff
@@ -26,10 +26,11 @@ statement
   : ALTER TABLE multipartIdentifier SET POLICY '(' retentionPolicy (columnRetentionPolicy)? ')'        #setRetentionPolicy
   | ALTER TABLE multipartIdentifier SET POLICY '(' replicationPolicy ')'                               #setReplicationPolicy
   | ALTER TABLE multipartIdentifier SET POLICY '(' sharingPolicy ')'                                   #setSharingPolicy
+  | ALTER TABLE multipartIdentifier SET POLICY '(' historyPolicy ')'                                   #setHistoryPolicy
   | ALTER TABLE multipartIdentifier MODIFY columnNameClause SET columnPolicy                           #setColumnPolicyTag
   | GRANT privilege ON grantableResource TO principal                                                  #grantStatement
   | REVOKE privilege ON grantableResource FROM principal                                               #revokeStatement
-  | SHOW GRANTS ON grantableResource       #showGrantsStatement
+  | SHOW GRANTS ON grantableResource                                                                   #showGrantsStatement
```
Comment (maluchari) [inline integrations/spark/spark-3.1/openhouse-spark-runtime/src/main/antlr/com/linkedin/openhouse/spark/sql/catalyst/parser/extensions/OpenhouseSqlExtensions.g4:33]: alignment only change?

File: integrations/spark/spark-3.1/openhouse-spark-runtime/src/main/scala/com/linkedin/openhouse/spark/sql/catalyst/parser/extensions/OpenhouseSqlExtensionsAstBuilder.scala [hunk 5]
```diff
@@ -19,10 +19,11 @@ class OpenhouseSqlExtensionsAstBuilder (delegate: ParserInterface) extends Openh
 
   override def visitSetRetentionPolicy(ctx: SetRetentionPolicyContext): SetRetentionPolicy = {
     val tableName = typedVisit[Seq[String]](ctx.multipartIdentifier)
-    val (granularity, count) = typedVisit[(String, Int)](ctx.retentionPolicy())
+    val retentionPolicy = ctx.retentionPolicy()
```
Comment (maluchari) [inline integrations/spark/spark-3.1/openhouse-spark-runtime/src/main/scala/com/linkedin/openhouse/spark/sql/catalyst/parser/extensions/OpenhouseSqlExtensionsAstBuilder.scala:22]: why is this changed required as part of this PR?

**Review by maluchari — DISMISSED** [review]
minor comments. But overall LGTM

File: integrations/spark/spark-3.1/openhouse-spark-runtime/src/main/scala/com/linkedin/openhouse/spark/sql/catalyst/parser/extensions/OpenhouseSqlExtensionsAstBuilder.scala [hunk 6]
```diff
@@ -158,13 +159,34 @@ class OpenhouseSqlExtensionsAstBuilder (delegate: ParserInterface) extends Openh
...
+  }
+  override def visitHistoryPolicy(ctx: HistoryPolicyContext): (Option[String], Int, Int) = {
+    val maxAgePolicy = if (ctx.maxAge() != null)
+        typedVisit[(String, Int)](ctx.maxAge().duration())
+      else (null, -1)
+    val versionPolicy = if (ctx.versions() != null)
+        typedVisit[Int](ctx.versions())
+      else -1
+    if (maxAgePolicy._2 == -1 && versionPolicy == -1) {
+      throw new OpenhouseParseException("Either MAX_AGE or VERSIONS must be specified in HISTORY policy", ctx.start.getLine, ctx.start.getCharPositionInLine)
```
Comment (abhisheknath2011) [inline integrations/spark/spark-3.1/openhouse-spark-runtime/src/main/scala/com/linkedin/openhouse/spark/sql/catalyst/parser/extensions/OpenhouseSqlExtensionsAstBuilder.scala:179]: Do we want to be explicit here for the error message like snapshot history or snapshot itself is replaced by history as I am seeing the the sql syntax.

**Review by abhisheknath2011 — DISMISSED** [review]
Thanks for refactoring the tests and answering the questions. Added minor comment, otherwise LGTM!

**Review by autumnust — DISMISSED** [review]
lgtm

File: integrations/spark/spark-3.1/openhouse-spark-runtime/src/main/scala/com/linkedin/openhouse/spark/sql/catalyst/parser/extensions/OpenhouseSqlExtensionsAstBuilder.scala (hunk 5 above)
Comment (Will-Lo) [reply integrations/spark/spark-3.1/openhouse-spark-runtime/src/main/scala/com/linkedin/openhouse/spark/sql/catalyst/parser/extensions/OpenhouseSqlExtensionsAstBuilder.scala:22]: Good catch I'll revert this change

File: integrations/spark/spark-3.1/openhouse-spark-runtime/src/main/scala/com/linkedin/openhouse/spark/sql/catalyst/parser/extensions/OpenhouseSqlExtensionsAstBuilder.scala (hunk 6 above)
Comment (Will-Lo) [reply integrations/spark/spark-3.1/openhouse-spark-runtime/src/main/scala/com/linkedin/openhouse/spark/sql/catalyst/parser/extensions/OpenhouseSqlExtensionsAstBuilder.scala:179]: I'll make it explicit, yes we use HISTORY as the syntax now
```
At least one of MAX_AGE or VERSIONS must be specified in HISTORY policy, e.g. " +
        "ALTER TABLE openhouse.db.table SET POLICY (HISTORY MAX_AGE=2D) or ALTER TABLE openhouse.db.table SET POLICY (HISTORY VERSIONS=3)
```

**Review by abhisheknath2011 — APPROVED** [review]
